## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics

df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [5]:
df['last'].unique()

array(['Harding', 'Myers', 'Watson', 'Baker', 'Schwartz', 'Reyes',
       'Moore', 'Wright', 'Brown', 'Garcia', 'Williams', 'Howard',
       'Gould', 'Taylor', 'Orozco', 'Jordan', 'Green', 'Stephenson',
       'Anderson', 'Roberts', 'Rhodes', 'Mendoza', 'Lowe', 'Rice', 'Cook',
       'Kennedy', 'Harris', 'Palmer', 'Robertson', 'Barnes', 'Wells',
       'Spencer', 'Brooks', 'Mitchell', 'Mccann', 'Bean', 'Griffith',
       'Wagner', 'Johnson', 'Newman', 'Hernandez', 'Davis', 'Simmons',
       'Turner', 'Aguilar', 'Kelly', 'Chase', 'Hobbs', 'Wilkinson',
       'Peterson', 'Cervantes', 'Carlson', 'Thompson', 'Vasquez', 'Ward',
       'Middleton', 'Arias', 'Ramirez', 'Pennington', 'Ellis', 'Mcmillan',
       'Morgan', 'Austin', 'Vincent', 'Campbell', 'Richardson', 'Colon',
       'Jackson', 'Espinoza', 'West', 'Schultz', 'Cannon', 'Hanson',
       'Gilbert', 'Powell', 'Lewis', 'Bennett', 'Evans', 'Martin',
       'Gross', 'Mullen', 'Hughes', 'Short', 'Castillo', 'Carr',
       'Farrell', 'C

In [6]:
df['job'].count()

54639

In [7]:
df.loc[df['unit_price'] < 1000, 'subcategory'].value_counts()

bathroom supplies     6318
paper products        5874
kitchen supplies      5633
tables                4540
misc                  4286
stands                4201
accounting            3756
project management    3522
plants                3466
wall art              2989
seating               2664
pens                  2603
logistics             1007
telecom                978
audio visual           732
cabling                655
desks                  510
computers              467
Name: subcategory, dtype: int64

In [8]:
# What three item categories had the most entries?

df['category'].value_counts().head(3)

consumables    23538
furniture      11915
software        8400
Name: category, dtype: int64

In [9]:
# For the category with the most entries, which subcategory had the most entries?

df.loc[df['category'] == 'consumables']['subcategory'].value_counts().head(1)

bathroom supplies    6424
Name: subcategory, dtype: int64

In [10]:
# Which five clients had the most entries in the data?

df['client_id'].value_counts().head()

33615    220
66037    211
46820    209
38378    207
24741    207
Name: client_id, dtype: int64

In [11]:
# Store the client ids of those top 5 clients in a list.

#list(df['client_id'].value_counts().head(5).index)

top_5_client_ids = [33615, 66037, 46820, 38378, 24741]
top_5_client_ids

[33615, 66037, 46820, 38378, 24741]

In [12]:
# How many total units (the qty column) did the client with the most entries order order?

df.loc[df['client_id'] == 33615, ['qty']].sum().head()['qty']

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [13]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty

df['line_subtotal'] = df['qty'] * df['unit_price']
totals = df[['unit_price', 'qty', 'line_subtotal']]
totals.head(2)

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [14]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

df['total_weight'] = df['unit_weight'] * df['qty']

def shipping_calculator(weight):
    if weight > 50:
        return 7 * weight
    else:
        return 10 *weight
    
df['total_weight'] = df['qty'] * df['unit_weight']
df['shipping_price'] = df['total_weight'].apply(shipping_calculator)
df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head(3)

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [15]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

tax = 0.0925
df['line_subtotal'] = df['unit_price'] * df['qty']

df['line_price'] = ((tax*df['shipping_price']) + (tax*df['line_subtotal']) +  df['line_subtotal'].round(2) + df['shipping_price']).round(2)
df[['line_subtotal', 'shipping_price', 'line_price']].head(3)

line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [16]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

df['line_cost'] = ((df['unit_cost']* df['qty']) + df['shipping_price']).round(2)
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [17]:
# Create a column for the profit of each line using line cost and line price

df['line_profit'] = (df['line_price'] - df['line_cost'])
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [18]:
# Check your work using the totals above

order_list = [2742071, 2173913, 6128929]
for order_id in order_list:
    check_totals = round(df.loc[df['order_id'] == order_id, 'line_price'].sum(), 2)
    print(f'Order {order_id} total: ${check_totals}')
    

Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [19]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

client_list = [33615, 66037, 46820, 38378, 24741]
for client_id in client_list:
    check_totals = df.loc[df['client_id'] == client_id, 'line_price'].sum().round(2)
    print(f'{client_id} : ${check_totals}')


33615 : $8377308.52
66037 : $10259514.79
46820 : $9743794.36
38378 : $12906550.87
24741 : $82268892.02


In [20]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

s = df['client_id']
s.isin([33615, 66037, 46820, 38378, 24741])
top_df = df[s.isin([33615, 66037, 46820, 38378, 24741])].groupby(['client_id']).sum()
top_df = top_df[['qty', 'shipping_price', 'line_price', 'line_cost', 'line_profit']].loc[client_list].reset_index()
top_df

client_id     qty  shipping_price   line_price    line_cost  line_profit
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31

In [21]:
# Format the data and rename the columns to names suitable for presentation.

rename_dict = {
    'client_id': 'Client ID',
    'qty': 'Units',
    'shipping_price': 'Shipping (Millions)',
    'line_price': 'Total Revenue (Millions)',
    'line_cost': 'Total Cost (Millions)',
    'line_profit': 'Total Profit (Millions)'
}
top_df = top_df.rename(columns=rename_dict)


# Define the money columns. 
money_cols = ['Shipping (Millions)', 'Total Revenue (Millions)', 'Total Cost (Millions)', 'Total Profit (Millions)']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    return x/1000000
    

# Apply the currency_format_millions function to only the money columns. 
for col in money_cols:
    top_df[col] = top_df[col].apply(currency_format_millions)


# Rename the columns to reflect the change in the money format. 
top_df



Client ID   Units  Shipping (Millions)  Total Revenue (Millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (Millions)  Total Profit (Millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [22]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
high_to_low = top_df.sort_values('Total Profit (Millions)', ascending=False)
high_to_low

Client ID   Units  Shipping (Millions)  Total Revenue (Millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (Millions)  Total Profit (Millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995

In [23]:
mean = df['line_profit'].mean()
med = df['line_profit'].median()
s = df['line_profit'].sum()
top_100 = df.nlargest(100, 'line_profit')['line_profit'].sum()
print(df['client_id'].count())
print(med, f'{s:,.2f}', f'top 100: {top_100:,.2f}')
print(mean, f'{s:,.2f}',f'top 100: {top_100:,.2f}')

54639
1894.8400000000001 1,605,953,284.81 top 100: 863,032,691.36
29392.069489009664 1,605,953,284.81 top 100: 863,032,691.36


# Summary
I discovered the top client, Client ID 24741, accounted for \\$ 36.58 million in total profit, which was a lot more profit than the other four clients in the top five, and even more than all of them put together. 

Then looking at the data further, I found out the median for Total Profit was \\$ 1,894.84, which was drastically skewed from the top client's total profit of \\$ 36.58 million. The mean was \\$ 29,392, and also far way from the top client's total profit. Lastly the top 100 clients' total profit accounted for over half the total profit for all the clients put together. In closing, the clients at the top of our dataframe accounted for a lot more profit than the others, which caused the median and mean to be lower than if the total profits were distributed more evenly.